nitrite_test.go

```diff
@@ -7,79 +7,110 @@ import (
 	"testing"
 	"time"
 
-	"github.com/blocky/nitrite"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+
+	"github.com/blocky/nitrite"
 )
 
-// An easy way to create a new nitro attestation is to use delphi-cli.
-// Request an attestation over some endpoint and copy the resulting
-// "enclave_attestation" string from the output---it should be a
-// base64 encoded string
-//
 //go:embed testdata/nitro_attestation.txt
 var nitroAttestString string
 
 func TestNitrite_Verify(t *testing.T) {
 	attestBytes, err := base64.StdEncoding.DecodeString(nitroAttestString)
 	require.NoError(t, err)
-
-	doc, err := nitrite.NewDocumentFromCosePayloadBytes(attestBytes)
+	attestTime, err := nitrite.Timestamp(attestBytes)
 	require.NoError(t, err)
 
 	roots := x509.NewCertPool()
 	ok := roots.AppendCertsFromPEM(nitrite.AWSNitroEnclavesCertPEM)
 	require.True(t, ok)
 
 	t.Run("happy path", func(t *testing.T) {
-		opts := nitrite.VerifyOptions{
-			CurrentTime: doc.CreatedAt(),
-			Roots:       roots,
-		}
+		// when
+		result, err := nitrite.Verify(
+			attestBytes,
+			nitrite.WithDefaultRootCert(),
+			nitrite.WithAttestationTime(),
+		)
 
-		result, err := nitrite.Verify(attestBytes, opts)
+		// then
 		require.NoError(t, err)
-		assert.Equal(t, *result.Document, *doc)
-		assert.True(t, result.SignatureOK)
+		require.NotEmpty(t, result)
 	})
 
-	t.Run("certificate has expired or is not yet valid", func(t *testing.T) {
-		opts := nitrite.VerifyOptions{
-			CurrentTime: time.UnixMilli(0),
-			Roots:       roots,
-		}
+	unknownSigningAuthorityTests := []struct {
+		name        string
+		rootCertOpt nitrite.RootCertFunc
+	}{
+		{
+			"nil roots",
+			nitrite.WithRootCert(nil),
+		},
+		{
+			"empty roots",
+			nitrite.WithRootCert(x509.NewCertPool()),
+		},
+	}
+	for _, tt := range unknownSigningAuthorityTests {
+		t.Run(tt.name, func(t *testing.T) {
+			// when
+			_, err = nitrite.Verify(
+				attestBytes,
+				tt.rootCertOpt,
+				nitrite.WithAttestationTime(),
+			)
 
-		result, err := nitrite.Verify(attestBytes, opts)
-		assert.ErrorContains(t, err, "certificate has expired or is not yet valid")
-		assert.Nil(t, result)
-	})
-}
+			// then
+			assert.ErrorContains(t, err, "verifying certificate")
+		})
+	}
 
-func TestNitriteDocument_NewDocumentFromCosePayloadBytes(t *testing.T) {
-	attestBytes, err := base64.StdEncoding.DecodeString(nitroAttestString)
-	require.NoError(t, err)
+	timeOutOfBoundsTests := []struct {
```

Comment:
Nice this is clean! I didn't know about `.Add()`

Reply:
Thanks! Yes, convenient!